Add support for GS105PE

frank-f opened the pull request:
I stumbled across this library via Home Assistant. I have a GS105PE here which unfortunately is not supported.

I went through the code and patched my model in, which seems to use some kind of mixture between API v1 and v2.

lorenzo-deluca commented:
What a case 👍 
I was also working on the integration of the same model, i had it in my repository, i did PR after you. i leave it to the good @foxey to choose and merge 👍

foxey commented on src/py_netgear_plus/parsers.py:
```diff
@@ -230,12 +247,18 @@ def parse_port_status(
-            portstatus_elems = tree.xpath('//tr[@class="portID"]/td[3]')
-            portspeed_elems = tree.xpath('//tr[@class="portID"]/td[4]')
-            portconnectionspeed_elems = tree.xpath('//tr[@class="portID"]/td[5]')
+            if get_text_from_next_element(tree, '//td[text()="Port"]') == "Port Description":
```
Does this also work for German language settings? "Port Description" will probably not the same.

Also, the maintainability of the code doesn't improve with an additional if statement. Can't we better overload the parse_port_status method in the GS105PE class definition?

frank-f replied:
No objections to duplicating the whole function. I was trying to keep code redundancy to a minimum, but using a dedicated function will have the benefit of working for all languages, since we can use static indexes.

foxey replied:
I get it. I had a lot of problems with regression in other models when working on supporting additional ones. I think the duplication is worth the reduction of regression.

foxey commented on src/py_netgear_plus/parsers.py:
```diff
@@ -318,6 +343,32 @@ def parse_port_statistics_v2(
+        crc_current_elems = tree.xpath('//tr[@class="portID"]/input[6]')
+
+        # calculate int bytes
+        rx = [turnover*4294967296+current for turnover,current in zip(convert_to_int(rx_turnover_elems, output_elems=ports, base=10, attr_name="value"), convert_to_int(rx_current_elems, output_elems=ports, base=10, attr_name="value"))]
```
These long lines cause lint errors. Please split them up.

foxey commented on src/py_netgear_plus/parsers.py:
```diff
@@ -318,6 +343,32 @@ def parse_port_statistics_v2(
             "speed_io": io_zeros,
         }
 
+    def parse_port_statistics_v2b(
```
Same here, the v2b version is the only one that you are using for the GS105PE model, are you?
In that case, we can better override the `parse_port_statistics` method in the GS105PE class definition.

foxey commented on src/py_netgear_plus/parsers.py:
```diff
@@ -351,6 +402,39 @@ def __init__(self) -> None:
+
+        switch_name = get_first_value(tree, '//input[@id="switch_name"]')
+        switch_serial_number = get_text_from_next_element(
+            tree, '//td[contains(text(),"Serial Number")]'
```
This probably doesn't work in German language mode. Is it possible to change the xpath expression to something language independent?

frank-f replied:
This is going to be difficult. There is no unique identifier for any of the fields on the switch info page. The only other approach I could think of is counting the rows and using an n-th field to set the values.

foxey replied:
Okay, let's go with this and if someone in DE opens an issue, we'll fix it then.

foxey reviewed (changes requested):
Great to see others step in and contribute! I've added a couple of comments and proposed changes. Also, it would be good to add the model to the unit tests, by adding html pages in the the `pages/` folder (@lorenzo-deluca already did that, but did not update the test definitions).

foxey commented:
@frank-f, @lorenzo-deluca maybe the two of you can agree on who's updating his PR? I don't want you to duplicate more effort!

frank-f commented:
> @frank-f, @lorenzo-deluca maybe the two of you can agree on who's updating his PR? I don't want you to duplicate more effort!

I'll push some of the requested changes in a few minutes, but if you @lorenzo-deluca like some of your ideas better than mine, feel free to add something to this PR as well.

frank-f commented on src/py_netgear_plus/models.py:
```diff
@@ -58,6 +58,11 @@ class AutodetectedSwitchModel:
         }
     ]
     PORT_STATUS_TEMPLATES: ClassVar = [
+        {
```
GET is correct.

frank-f commented:
> Great to see others step in and contribute! I've added a couple of comments and proposed changes. Also, it would be good to add the model to the unit tests, by adding html pages in the the `pages/` folder (@lorenzo-deluca already did that, but did not update the test definitions).

I added @lorenzo-deluca's pages to this PR, but I am not sure what you mean by test definitions. What do I need to add where?

foxey commented on src/py_netgear_plus/models.py:
```diff
@@ -129,6 +124,12 @@ class GS105PE(AutodetectedSwitchModel):
         ("check_login_form_rand", [True]),
         ("parse_login_title_tag", ["GS105PE"]),
     ]
+    PORT_STATUS_TEMPLATES: ClassVar = [
```
If this is the correct one, don't we need to correct them as well for `PORT_STATISTICS_TEMPLATES` as well? It currently inherits the POST request from the parent class.

frank-f replied:
Yes, indeed, it is a GET request as well. I added it.

foxey reviewed (changes requested):
Thanks for following up so quickly! I think we're almost there.
Can you also run the `scripts/lint` script to fix the linting errors?

foxey commented:
> > Great to see others step in and contribute! I've added a couple of comments and proposed changes. Also, it would be good to add the model to the unit tests, by adding html pages in the the `pages/` folder (@lorenzo-deluca already did that, but did not update the test definitions).
> 
> I added @lorenzo-deluca's pages to this PR, but I am not sure what you mean by test definitions. What do I need to add where?

@frank-f I run a series of pytest unit tests (in `tests/test__init__.py`). Some changes are needed to include this model in the tests as well, but let's first merge you're PR and I'll add the model to the test suite.

To be able to do that, I also need a copy of the login page (probably `login.cgi`). It's in the `pages` folder, but it's an empty file.

foxey commented:
Hi @frank-f, I was adding the unit tests, but I think the page copy for the `SWITCH_INFO_TEMPLATES` is missing (probably `switch_info.cgi`). Can you add a copy of this page?

frank-f commented:
> Hi @frank-f, I was adding the unit tests, but I think the page copy for the `SWITCH_INFO_TEMPLATES` is missing (probably `switch_info.cgi`). Can you add a copy of this page?

Done

foxey reviewed (approved):
Thanks! I've merged your changes with the branch I'm working on.
The only change I've still made is the `strict` option in the `zip` call. The number of elements in turnover and current should be equal, isn't it?
In that case using `strict=True` will catch any mistakes where the number of elements in both array is not the same.

frank-f commented:
> Thanks! I've merged your changes with the branch I'm working on. The only change I've still made is the `strict` option in the `zip` call. The number of elements in turnover and current should be equal, isn't it? In that case using `strict=True` will catch any mistakes where the number of elements in both array is not the same.

I think `strict=True` is not needed, because `convert_to_int()` takes care that the resulting number of elements will always be `output_elems`, but setting it will surely not hurt anyone either.
